src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,601 @@
+If line 3 cares only about strings,
+then you can use `String` in yout type check.
+Because `String` is a subtype of `Comparable`, the promotion works:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+void f(Object o) {
+  if (o is Comparable) {              // (1)
+    if (o is [!String!]) {               // (2)
+      print(o.matchAsPrefix('foo')); // (3) OK; String is a subtype of Comparable.
+    }
+  }
+}
+{% endprettify %}
+
+
+### Write captured by a local function
+
+**The cause:**
+The variable has been write captured by
+a local function or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  var foo = () {
+    i = j;
+  };
+  // ... Use foo ... 
+  if (i == null) return; // (1)
+  // ... Additional code ...
+  print(i.isEven);       // (2) ERROR
+}
+{% endprettify %}
+
+Flow analysis reasons that as soon as the definition of `foo` is reached,
+it might get called at any time,
+therefore it's no longer safe to promote `i` at all.
+As with loops, this demotion happens regardless of
+the type of the right hand side of the assignment.
+
+Sometimes it's possible to restructure the logic so that
+the promotion is before the write capture:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return; // (1)
+  // ... Additional code ...
+  print(i.isEven);       // (2) OK
+[!  var foo = () {
+    i = j;
+  };
+  // ... Use foo ...!] 
+}
+{% endprettify %}
+
+Another option is to use a separate variable that isn't write captured:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  var foo = () {
+    i = j;
+  };
+  // ... Use foo ... 
+  [!var i2 = i;!]
+  if ([!i2!] == null) return; // (1)
+  // ... Additional code ...
+  print([!i2!].isEven);       // (2) OK because `i2` isn't write captured.
+}
+{% endprettify %}
+
+Or you can do a redundant check:
+
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  var foo = () {
+    i = j;
+  };
+  // ... Use foo ... 
+  if (i == null) return; // (1)
+  // ... Additional code ...
+  print(i[!!!].isEven);     // (2) OK due to `!` check.
+}
+{% endprettify %}
+
+
+### Written outside of the current closure or function expression {#write-outer}
+
+**The cause:**
+The variable is written to outside of a closure or function expression,
+and the type promotion location is
+inside the closure or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return;
+  var foo = () {
+    print(i.isEven); // (1) ERROR
+  };
+  i = j;             // (2)
+}
+{% endprettify %}
+
+Flow analysis reasons that there's no way to determine
+when `foo` might get called,
+so it might get called after the assignment at (2),
+and thus the promotion might no longer be valid.
+As with loops, this demotion happens regardless of the type of
+the right hand side of the assignment.
+
+A solution is to use a fresh variable:
```

Comment:
I'll propose uniformly using a phase such as 'create a new variable', 'add a new variable', or even 'define a new variable' when a new variable is being introduced to solve the problem. For example: 'A solution is to create a new variable:'. Adding 'local' would also be reasonable: 'A solution is to create a new local variable:'.

Reply:
I like the more specific one: "create a new local variable". Thanks, @bwilkerson!

Reply:
Upon further thought, the "new" seems redundant, so I'm thinking "create a local variable".

Reply:
I agree. sgtm